DevTools/HackStudio/Git/GitRepo.cpp

```diff
@@ -0,0 +1,196 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "GitRepo.h"
+#include <AK/ScopeGuard.h>
+#include <LibCore/File.h>
+#include <fcntl.h>
+#include <spawn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/wait.h>
+
+namespace HackStudio {
+
+GitRepo::CreateResult GitRepo::try_to_create(const LexicalPath& repository_root)
+{
+    if (!git_is_installed()) {
+        return { CreateResult::Type::GitProgramNotFound, nullptr };
+    }
+    if (!git_repo_exists(repository_root)) {
+        return { CreateResult::Type::NoGitRepo, nullptr };
+    }
+
+    return { CreateResult::Type::Success, adopt(*new GitRepo(repository_root)) };
+}
+
+RefPtr<GitRepo> GitRepo::initialize_repository(const LexicalPath& repository_root)
+{
+    auto res = command_wrapper("init", repository_root);
+    if (!res.has_value())
+        return {};
+
+    ASSERT(git_repo_exists(repository_root));
+    return adopt(*new GitRepo(repository_root));
+}
+
+Vector<LexicalPath> GitRepo::unstaged_files() const
+{
+    auto modified = modified_files();
+    auto untracked = untracked_files();
+    modified.append(untracked.data(), untracked.size());
+    return modified;
+}
+
+Vector<LexicalPath> GitRepo::staged_files() const
+{
+    auto raw_result = command("diff --cached --name-only");
+    if (!raw_result.has_value())
+        return {};
+    return parse_files_list(raw_result.value());
+}
+
+Vector<LexicalPath> GitRepo::modified_files() const
+{
+    auto raw_result = command("ls-files --modified --exclude-standard");
+    if (!raw_result.has_value())
+        return {};
+    return parse_files_list(raw_result.value());
+}
+
+Vector<LexicalPath> GitRepo::untracked_files() const
+{
+    auto raw_result = command("ls-files --others --exclude-standard");
+    if (!raw_result.has_value())
+        return {};
+    return parse_files_list(raw_result.value());
+}
+
+Vector<LexicalPath> GitRepo::parse_files_list(const String& raw_result)
+{
+    auto lines = raw_result.split('\n');
+    Vector<LexicalPath> files;
+    for (const auto& line : lines) {
+        files.empend(line);
+    }
+    return files;
+}
+
+Optional<String> GitRepo::command(const String& git_command) const
```

Comment:
`String` already has a null state, so let's use that instead of wrapping it in `Optional` :)